File: KeyStats.Windows/KeyStats/Services/DisplayStatsAggregator.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using KeyStats.Models;

namespace KeyStats.Services;

public sealed class DisplayStatsAggregator
{
    private readonly RemoteShardCache _remoteCache;
    private readonly Func<string> _localDeviceId;

    public DisplayStatsAggregator(RemoteShardCache remoteCache, Func<string> localDeviceId)
    {
        _remoteCache = remoteCache;
        _localDeviceId = localDeviceId;
    }

    public DailyStats Aggregate(DateTime date, DailyStats localStats)
    {
        var day = date.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture);
        var result = Clone(localStats, date.Date);
        var localDeviceId = _localDeviceId();

        foreach (var record in _remoteCache.GetAll().Where(record =>
                     string.Equals(record.Plaintext.LocalDay, day, StringComparison.Ordinal) &&
                     !string.Equals(record.DeviceId, localDeviceId, StringComparison.Ordinal)))
        {
            var remote = record.Plaintext;
            result.KeyPresses = SafeAdd(result.KeyPresses, remote.KeyPresses);
            result.LeftClicks = SafeAdd(result.LeftClicks, remote.Clicks.Left);
            result.RightClicks = SafeAdd(result.RightClicks, remote.Clicks.Right);
            result.MiddleClicks = SafeAdd(result.MiddleClicks, remote.Clicks.Middle);
            result.SideBackClicks = SafeAdd(result.SideBackClicks, remote.Clicks.SideBack);
            result.SideForwardClicks = SafeAdd(result.SideForwardClicks, remote.Clicks.SideForward);

            foreach (var keyCount in remote.KeyPressCounts)
            {
                var key = (keyCount.Key ?? string.Empty).Trim();
                if (string.IsNullOrWhiteSpace(key) || keyCount.Value <= 0) continue;
                result.KeyPressCounts[key] = SafeAdd(
                    result.KeyPressCounts.TryGetValue(key, out var existing) ? existing : 0,
                    keyCount.Value);
            }
        }

        return result;
    }

    public IReadOnlyCollection<DateTime> GetRemoteDays()
    {
        return _remoteCache.GetAvailableDays()
            .Select(day => DateTime.TryParseExact(
                day,
                "yyyy-MM-dd",
                CultureInfo.InvariantCulture,
                DateTimeStyles.None,
                out var parsed)
                ? (DateTime?)parsed.Date
                : null)
            .Where(value => value.HasValue)
            .Select(value => value!.Value)
            .ToList();
    }

    private static DailyStats Clone(DailyStats source, DateTime date)
    {
        return new DailyStats(date)
        {
            KeyPresses = Math.Max(0, source.KeyPresses),
            KeyPressCounts = new Dictionary<string, int>(source.KeyPressCounts, StringComparer.Ordinal),
            LeftClicks = Math.Max(0, source.LeftClicks),
            RightClicks = Math.Max(0, source.RightClicks),
            MiddleClicks = Math.Max(0, source.MiddleClicks),
            SideBackClicks = Math.Max(0, source.SideBackClicks),
            SideForwardClicks = Math.Max(0, source.SideForwardClicks),
            MouseDistance = Math.Max(0, source.MouseDistance),
            ScrollDistance = Math.Max(0, source.ScrollDistance),
            PeakKPS = Math.Max(0, source.PeakKPS),
            PeakCPS = Math.Max(0, source.PeakCPS),
            AppStats = source.AppStats.ToDictionary(pair => pair.Key, pair => new AppStats(pair.Value))
        };
    }

    private static int SafeAdd(int left, long right)
    {
        if (right <= 0) return Math.Max(0, left);
        var total = (long)Math.Max(0, left) + right;
        return total >= int.MaxValue ? int.MaxValue : (int)total;
    }
}

File: KeyStats/Sync/DisplayStatsAggregator.swift
import Foundation

enum DisplayStatsAggregator {
    static func coreSnapshot(
        from stats: DailyStats,
        deviceId: String,
        revision: Int64
    ) throws -> CoreDaySnapshotV1 {
        var counts: [String: Int64] = [:]
        for (rawKey, rawCount) in stats.keyPressCounts {
            let key = SyncKeyCanonicalizer.canonicalize(rawKey, platform: "mac")
            guard !key.isEmpty else { continue }
            let count = Int64(max(0, rawCount))
            counts[key] = SyncMath.saturatingAdd(counts[key] ?? 0, count)
        }
        return try CoreDaySnapshotV1(
            deviceId: deviceId,
            localDay: SyncDay.string(from: stats.date),
            revision: revision,
            keyPresses: Int64(max(0, stats.keyPresses)),
            keyPressCounts: counts,
            clicks: CoreClickSnapshotV1(
                left: Int64(max(0, stats.leftClicks)),
                right: Int64(max(0, stats.rightClicks)),
                middle: Int64(max(0, stats.middleClicks)),
                sideBack: Int64(max(0, stats.sideBackClicks)),
                sideForward: Int64(max(0, stats.sideForwardClicks))
            )
        ).validated()
    }

    static func aggregate(
        local: [String: DailyStats],
        remote: [CoreDaySnapshotV1],
        currentDeviceId: String
    ) -> [String: DailyStats] {
        var result = local
        for shard in remote where shard.deviceId != currentDeviceId {
            guard let date = SyncDay.date(from: shard.localDay) else { continue }
            var day = result[shard.localDay] ?? DailyStats(date: date)
            day.keyPresses = add(day.keyPresses, shard.keyPresses)
            day.leftClicks = add(day.leftClicks, shard.clicks.left)
            day.rightClicks = add(day.rightClicks, shard.clicks.right)
            day.middleClicks = add(day.middleClicks, shard.clicks.middle)
            day.sideBackClicks = add(day.sideBackClicks, shard.clicks.sideBack)
            day.sideForwardClicks = add(day.sideForwardClicks, shard.clicks.sideForward)
            for (key, count) in shard.keyPressCounts {
                day.keyPressCounts[key] = add(day.keyPressCounts[key] ?? 0, count)
            }
            result[shard.localDay] = day
        }
        return result
    }

    static func currentDay(
        local: DailyStats,
        remote: [CoreDaySnapshotV1],
        currentDeviceId: String
    ) -> DailyStats {
        let day = SyncDay.string(from: local.date)
        let remoteForDay = deduplicatedLatest(remote.filter { $0.localDay == day })
        return aggregate(
            local: [day: local],
            remote: remoteForDay,
            currentDeviceId: currentDeviceId
        )[day] ?? local
    }

    static func deduplicatedLatest(_ snapshots: [CoreDaySnapshotV1]) -> [CoreDaySnapshotV1] {
        var latest: [String: CoreDaySnapshotV1] = [:]
        for snapshot in snapshots {
            let key = "\(snapshot.deviceId)|\(snapshot.localDay)"
            guard let existing = latest[key] else {
                latest[key] = snapshot
                continue
            }
            if snapshot.revision > existing.revision {
                latest[key] = snapshot
            }
        }
        return Array(latest.values)
    }

    private static func add(_ local: Int, _ remote: Int64) -> Int {
        guard remote > 0 else { return max(0, local) }
        if remote >= Int64(Int.max) { return Int.max }
        return SyncMath.saturatingAdd(max(0, local), Int(remote))
    }
}
